### PR TITLE
NLL diagnostics replaced nice closure errors w/ indecipherable free region errors 

### LINKED ISSUE #51027: NLL diagnostics replaced nice closure errors w/ indecipherable free region errors
The following tests all have a common weakness around how they report errors under NLL. Under AST borrowck they would say something somewhat understandable about closures. But under NLL they don't anymore.

List:

 * [ ] borrowck/issue-45983.rs
 * [ ] borrowck/issue-7573.rs
 * [ ] borrowck/regions-escape-bound-fn-2.rs
 * [ ] borrowck/regions-escape-bound-fn.rs
 * [ ] borrowck/regions-escape-unboxed-closure.rs
 * [ ] closure-expected-type/expect-region-supply-region.rs
 * [ ] error-codes/E0621-does-not-trigger-for-closures.rs

(This list of tests is drawn from an [informal paper](https://paper.dropbox.com/doc/NLL-stderr-diagnostic-deviations-KzroSvNgQ7lip3WQyiLTa) document that I have been using to keep notes for myself as I work on this...)

Example (from [`ui/borrowck/issue-45983.rs`](https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/issue-45983.rs)): 

AST borrowck says: 

```
error: borrowed data cannot be stored outside of its closure
  --> $DIR/issue-45983.rs:17:27
   |
LL |     let x = None;
   |         - borrowed data cannot be stored into here...
LL |     give_any(|y| x = Some(y));
   |              ---          ^ cannot be stored outside of its closure
   |              |
| ...because it cannot outlive this closure
```

NLL says:

```
error: free region `` does not outlive free region `'_#2r`
  --> $DIR/issue-45983.rs:17:27
   |
LL |     give_any(|y| x = Some(y));
   |                           ^
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #51027.

r? @nikomatsakis

### CODE REVIEW

**Comment by davidtwco** [comment]
As of submission, this PR produces the following error (or equivalent):

```
error: borrowed data escapes outside of closure
  --> issue-45983.rs:17:27
   |
16 |     let mut x = None;
   |         ----- `x` is declared here, outside of the closure body
17 |     give_any(|y| x = Some(y));:
   |               -  ^^^^^^^^^^^ `y` escapes the closure body here
   |               |
   |               `y` is a reference that is only valid in the closure body
```

It fixes the following tests mentioned in the issue
- [x] borrowck/issue-45983.rs
- [ ] borrowck/issue-7573.rs (isn't an assigment, I've experimented here briefly but not got anywhere)
- [x] borrowck/regions-escape-bound-fn-2.rs
- [x] borrowck/regions-escape-bound-fn.rs
- [x] borrowck/regions-escape-unboxed-closure.rs
- [x] closure-expected-type/expect-region-supply-region.rs
- [ ] error-codes/E0621-does-not-trigger-for-closures.rs (not sure how to approach)

File: src/librustc/mir/tcx.rs [hunk 1]
```diff
@@ -119,6 +119,39 @@ impl<'tcx> Place<'tcx> {
...
+                        let base_ty = proj.base.ty(mir, *tcx).to_ty(*tcx);
+
+                        if  base_ty.is_closure() || base_ty.is_generator() {
+                            Some(field)
+                        } else {
+                            None
+                        }
+                    },
+                    ProjectionElem::Deref => {
+                        if recurse {
```
Comment (davidtwco) [inline src/librustc/mir/tcx.rs]: I wasn't sure about adding this parameter - without it, a handful of unrelated tests start throwing duplicate errors with shorter spans.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        debug!("get_var_name_and_span_for_region: attempting upvar");
+        self.get_upvar_index_for_region(tcx, fr)
+            .map(|index| self.get_upvar_name_and_span_for_region(tcx, mir, index))
+            .or_else(|| {
+                debug!("get_var_name_and_span_for_region: attempting argument");
+                self.get_argument_index_for_region(tcx, fr)
+                    .map(|index| self.get_argument_name_and_span_for_region(mir, index))
+            })
+            .unwrap_or_else(|| span_bug!(mir.span, "can't find var name for free region {:?}", fr))
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I'm a bit nervous about this case. Based on the other code I see, it seems like it is possible for the region to appear in the closure output, for example. I'll try to build your branch and produce an example though because it's hard for me to predict in advance.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        let upvar_name = tcx.hir.name(upvar_node_id);
+        let upvar_span = tcx.hir.span(upvar_node_id);
+        debug!("get_upvar_name_and_span_for_region: upvar_name={:?} upvar_span={:?}",
+               upvar_name, upvar_span);
+
+        (Some(upvar_name), upvar_span)
+    }
+
+    /// Get argument index for a region.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I would word this as:

---

Search the argument types for one that references `fr` (which should be a free region). Returns `Some(_)` with the index of the input if one is found.

NB: In the case of a clsoure, the `index` is indexing into the signature as seen by the user--- in particular, index 0 is *not* the implicit `self` parameter.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        debug!(
+            "get_upvar_index_for_region: found {:?} in upvar {} which has type {:?}",
+            fr, upvar_index, upvar_ty,
+        );
+
+        Some(upvar_index)
+    }
+
+    /// Get upvar name and span for a region.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I would say something like:

---

Given the index of an upvar, finds its name (if any) and the span from where it was declared.

---

Question: Why does this return an `Option`? From what I can tell, it is always `Some`?

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            .map(|index| self.get_upvar_name_and_span_for_region(tcx, mir, index))
+            .or_else(|| {
+                debug!("get_var_name_and_span_for_region: attempting argument");
+                self.get_argument_index_for_region(tcx, fr)
+                    .map(|index| self.get_argument_name_and_span_for_region(mir, index))
+            })
+            .unwrap_or_else(|| span_bug!(mir.span, "can't find var name for free region {:?}", fr))
+    }
+
+    /// Get upvar index for a region.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I would say something like:

---

Search the upvars (if any) to find one that references `fr`. Return its index.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+        debug!(
+            "get_argument_index_for_region: found {:?} in argument {} which has type {:?}",
+            fr, argument_index, self.universal_regions.unnormalized_input_tys[argument_index],
+        );
+
+        Some(argument_index)
+    }
+
+    /// Get argument name and span for a region.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I would reword this similar to the fn for upvars.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs [hunk 7]
```diff
@@ -224,30 +235,75 @@ impl<'tcx> RegionInferenceContext<'tcx> {
 
         // Get a span
         let (category, span) = categorized_path.first().unwrap();
+
+        match category {
+            ConstraintCategory::AssignmentToUpvar =>
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: This is interesting. Looking for an assignment to an upvar is not quite taking the approach I expected, but it also makes sense. I guess though that this code won't fire for something like:

```rust
fn foo(x: &mut Vec<&u32>) {
  bar(|y| x.push(y));
}

fn bar(_: impl FnOnce(&u32)) { }
```

Right?

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: Seems ok, though I can think we can address this case via a similar set of errors, if we generalize the conditions somewhat away from *assignment* and more into looking at the kinds of regions involved. (Or maybe this is just another sort of case to specialize)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs (hunk 7 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: This is similar to the `issue-7573.rs` case that isn't yet handled. I messed around with making another variant for `ConstraintCategory` that would go through to the `report_closure_error` branch.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 4 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: I had it return an option so that the types where the same in the `and_then` sections above. I'll change this to use a `.map(...)` in that function rather than change the return type of this one.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 3 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Pushed a commit that addresses this.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 4 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Pushed a commit that addresses this.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 5 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Pushed a commit that addresses this.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 6 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Pushed a commit that addresses this.

File: src/librustc/mir/tcx.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc/mir/tcx.rs]: I've just checked and it seems like only doing a single deref still exhibits this behaviour.

Comment (davidtwco) [reply src/librustc/mir/tcx.rs]: Pushed a fix for this.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Most recent pushed commit introduces changes that ICE due to this in some tests.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs (hunk 7 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/mod.rs]: Fixed with most recent commit.

File: src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/region_infer/error_reporting/var_name.rs]: Another commit now stops those ICEs by falling back to previous error handling.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
Giving high priority, as improving this case is a EP2 blocker.
